### ISSUE TITLE
Problem with new module resolving

### ISSUE THREAD

**Issue by EloB** [expo/router#786]
### Which package manager are you using? (Yarn is recommended)

npm

### Summary

After the 2.0.0 release of `expo-router` than `colord` package stopped working. I'm not sure if it has something todo with the new module resolver or if `colord` is setup wrong but it's always been working.

https://github.com/omgovich/colord
https://unpkg.com/browse/colord@2.9.3/

I was able to get it to run by explicit using with file plus extension (like below) or going into `node_modules` of `colord` and remove `module` key in `package.json`.

```js
import { colord } from "colord/index.js";
```

Does other packages with `module` and `.mjs` files works?

![image](https://github.com/expo/router/assets/476567/5eb18658-8af0-47d5-aa5a-e6f9d671b7c5)

<details>
  <summary>Click here to show logs</summary>

```
❯ node -v && npm -v
v16.15.1
8.11.0

~/Documents/expo-router
❯ npx create-expo-app@latest --template tabs@49
✔ What is your app named? … my-app
✔ Downloaded and extracted project files.
> npm install
npm WARN deprecated @npmcli/move-file@1.1.2: This functionality has been moved to @npmcli/fs
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated uglify-es@3.3.9: support for ECMAScript is superseded by `uglify-js` as of v3.13.0

added 1522 packages, and audited 1523 packages in 17s

87 packages are looking for funding
  run `npm fund` for details

7 moderate severity vulnerabilities

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.


✅ Your project is ready!

To run your project, navigate to the directory and run one of the following npm commands.

- cd my-app
- npm run android
- npm run ios
- npm run web

~/Documents/expo-router [⏱ 20s]
❯ cd my-app/

my-app [ main][ v16.15.1]
❯ npm install colord

added 1 package, and audited 1524 packages in 963ms

87 packages are looking for funding
  run `npm fund` for details

7 moderate severity vulnerabilities

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.

my-app [ main][!][ v16.15.1]
❯ npm run web

> my-app@1.0.0 web
> expo start --web

Starting project at ~/Documents/expo-router/my-app
Experimental path aliases feature is enabled. Learn more
Starting Metro Bundler
▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄▄
█ ▄▄▄▄▄ █▀▄█▀ ██  █ ▄▄▄▄▄ █
█ █   █ █▄   ▄█▄▄██ █   █ █
█ █▄▄▄█ █ ▀█▀█▄█ ██ █▄▄▄█ █
█▄▄▄▄▄▄▄█ ▀▄█ █▄▀ █▄▄▄▄▄▄▄█
█ ▄▀ ▀█▄▄▀█ ▀█▄▀██▀  ▄▀▄▄▀█
█ ▄▀▀▀▄▄▄ ▄▄██ ▄▀ ▄▀ ▀▀█▄▄█
██ ▀▀▀█▄▀▀█▀▀   █▀█ ▄█ ██▀█
█▄▀██▄▀▄ ▄█▀█  ██ ▄▄ ▀▀██▄█
█▄▄█▄▄▄▄▄▀ ▄▄▄█▄  ▄▄▄ █ ▄ █
█ ▄▄▄▄▄ █▄▀ ▀██▄  █▄█  ▀▄▄█
█ █   █ █▀█▄▄█ ▀▀▄ ▄▄ █▀▄██
█ █▄▄▄█ █▀▄▀█▄  █  █▄  ▄█▄█
█▄▄▄▄▄▄▄█▄▄▄▄█▄▄█▄███▄▄█▄▄█

› Metro waiting on exp://192.168.50.169:8081
› Scan the QR code above with Expo Go (Android) or the Camera app (iOS)

› Web is waiting on http://localhost:8081

› Using Expo Go
› Press s │ switch to development build

› Press a │ open Android
› Press i │ open iOS simulator
› Press w │ open web

› Press j │ open debugger
› Press r │ reload app
› Press m │ toggle menu
› Press o │ open project code in your editor

› Press ? │ show all commands

Logs for your project will appear below. Press Ctrl+C to exit.
Web Bundling complete 6810ms
Web Bundling complete 6894ms
"transform" style array value is deprecated. Use space-separated string functions, e.g., "scaleX(2) rotateX(15deg)".
Web Bundling complete 28ms
 LOG  [web] Logs will appear in the browser console
Web Bundling complete 68ms
Web Bundling complete 26ms
"transform" style array value is deprecated. Use space-separated string functions, e.g., "scaleX(2) rotateX(15deg)".
Web Bundling complete 23ms
 LOG  [web] Logs will appear in the browser console
Unable to resolve "colord" from "app/(tabs)/index.tsx"
Error: While trying to resolve module `colord` from file `~/Documents/expo-router/my-app/app/(tabs)/index.tsx`, the package `~/Documents/expo-router/my-app/node_modules/colord/package.json` was successfully found. However, this package itself specifies a `main` module field that could not be resolved (`~/Documents/expo-router/my-app/node_modules/colord/index.mjs`. Indeed, none of these files exist:

  * ~/Documents/expo-router/my-app/node_modules/colord/index.mjs(.web.ts|.ts|.web.tsx|.tsx|.web.js|.js|.web.jsx|.jsx|.web.json|.json|.web.cjs|.cjs|.web.scss|.scss|.web.sass|.sass|.web.css|.css)
  * ~/Documents/expo-router/my-app/node_modules/colord/index.mjs/index(.web.ts|.ts|.web.tsx|.tsx|.web.js|.js|.web.jsx|.jsx|.web.json|.json|.web.cjs|.cjs|.web.scss|.scss|.web.sass|.sass|.web.css|.css)
    at DependencyGraph.resolveDependency (~/Documents/expo-router/my-app/node_modules/metro/src/node-haste/DependencyGraph.js:289:17)
    at Object.resolve (~/Documents/expo-router/my-app/node_modules/metro/src/lib/transformHelpers.js:169:21)
    at Graph._resolveDependencies (~/Documents/expo-router/my-app/node_modules/metro/src/DeltaBundler/Graph.js:473:35)
    at Graph._processModule (~/Documents/expo-router/my-app/node_modules/metro/src/DeltaBundler/Graph.js:261:38)
    at async Graph._traverseDependenciesForSingleFile (~/Documents/expo-router/my-app/node_modules/metro/src/DeltaBundler/Graph.js:249:5)
    at async Graph.traverseDependencies (~/Documents/expo-router/my-app/node_modules/metro/src/DeltaBundler/Graph.js:157:9)
    at async DeltaCalculator._getChangedDependencies (~/Documents/expo-router/my-app/node_modules/metro/src/DeltaBundler/DeltaCalculator.js:281:42)
    at async DeltaCalculator.getDelta (~/Documents/expo-router/my-app/node_modules/metro/src/DeltaBundler/DeltaCalculator.js:112:16)
    at async DeltaBundler.getDelta (~/Documents/expo-router/my-app/node_modules/metro/src/DeltaBundler.js:67:12)
    at async IncrementalBundler.updateGraph (~/Documents/expo-router/my-app/node_modules/metro/src/IncrementalBundler.js:224:19)
```
</details>

### Minimal reproducible example

```bash
npx create-expo-app@latest --template tabs@49

cd my-app

cat <<EOF | git apply -
diff --git a/app/(tabs)/index.tsx b/app/(tabs)/index.tsx
index 6cbee6d..61e528a 100644
--- a/app/(tabs)/index.tsx
+++ b/app/(tabs)/index.tsx
@@ -2,11 +2,13 @@ import { StyleSheet } from 'react-native';
 
 import EditScreenInfo from '@/components/EditScreenInfo';
 import { Text, View } from '@/components/Themed';
+import { colord } from 'colord';
 
 export default function TabOneScreen() {
   return (
     <View style={styles.container}>
       <Text style={styles.title}>Tab One</Text>
+      <Text>{colord("#fff").toRgbString()}</Text>
       <View style={styles.separator} lightColor="#eee" darkColor="rgba(255,255,255,0.1)" />
       <EditScreenInfo path="app/(tabs)/index.tsx" />
     </View>
diff --git a/package.json b/package.json
index baf50ae..1eab27c 100644
--- a/package.json
+++ b/package.json
@@ -15,6 +15,7 @@
   "dependencies": {
     "@expo/vector-icons": "^13.0.0",
     "@react-navigation/native": "^6.0.2",
+    "colord": "^2.9.3",
     "expo": "~49.0.3",
     "expo-font": "~11.4.0",
     "expo-linking": "~5.0.2",
EOF

npm install
npm run web
```

**Comment by zRelux** [comment]
try loading mjs with a metro config change
```
const { getDefaultConfig } = require('expo/metro-config');

const defaultConfig = getDefaultConfig(__dirname);

defaultConfig.resolver.sourceExts.push('mjs');

module.exports = defaultConfig;
```

**Comment by EloB** [comment]
> ```js
> defaultConfig.resolver.sourceExts.push('mjs');
> ```

That worked. Maybe that should go into the default config? :)

**Comment by dalley8626** [comment]
I still got the same issue, i am using turbo-repo

metro.config.js
```
const { getDefaultConfig } = require("expo/metro-config");
const path = require("path");

// Find the project and workspace directories
const projectRoot = __dirname;
// This can be replaced with `find-yarn-workspace-root`
const workspaceRoot = path.resolve(projectRoot, "../..");

const config = getDefaultConfig(__dirname, {
  // Enable CSS support.
  isCSSEnabled: true,
});

// 1. Watch all files within the monorepo
config.watchFolders = [workspaceRoot];
// 2. Let Metro know where to resolve packages and in what order
config.resolver.nodeModulesPaths = [
  path.resolve(projectRoot, "node_modules"),
  path.resolve(workspaceRoot, "node_modules"),
];
// 3. Force Metro to resolve (sub)dependencies only from the `nodeModulesPaths`
config.resolver.disableHierarchicalLookup = true;

config.resolver.sourceExts.push("mjs");

module.exports = config;
```

error
```
Error: Unable to resolve module ./node_modules/expo-router/entry from /Users/senghuotlay/Klotti/klotti/apps/seller-app/.: 

None of these files exist:
  * node_modules/expo-router/entry(.native|.web.ts|.native.ts|.ts|.web.tsx|.native.tsx|.tsx|.web.js|.native.js|.js|.web.jsx|.native.jsx|.jsx|.web.json|.native.json|.json|.web.cjs|.native.cjs|.cjs|.web.scss|.native.scss|.scss|.web.sass|.native.sass|.sass|.web.css|.native.css|.css|.web.mjs|.native.mjs|.mjs)
  * node_modules/expo-router/entry/index(.native|.web.ts|.native.ts|.ts|.web.tsx|.native.tsx|.tsx|.web.js|.native.js|.js|.web.jsx|.native.jsx|.jsx|.web.json|.native.json|.json|.web.cjs|.native.cjs|.cjs|.web.scss|.native.scss|.scss|.web.sass|.native.sass|.sass|.web.css|.native.css|.css|.web.mjs|.native.mjs|.mjs)
    at ModuleResolver.resolveDependency (/Users/senghuotlay/Klotti/klotti/node_modules/metro/src/node-haste/DependencyGraph/ModuleResolution.js:137:15)
    at DependencyGraph.resolveDependency (/Users/senghuotlay/Klotti/klotti/node_modules/metro/src/node-haste/DependencyGraph.js:260:43)
    at /Users/senghuotlay/Klotti/klotti/node_modules/metro/src/lib/transformHelpers.js:177:21
    at Server._resolveRelativePath (/Users/senghuotlay/Klotti/klotti/node_modules/metro/src/Server.js:1183:12)
    at async Server.requestProcessor [as _processBundleRequest] (/Users/senghuotlay/Klotti/klotti/node_modules/metro/src/Server.js:475:37)
    at async Server._processRequest (/Users/senghuotlay/Klotti/klotti/node_modules/metro/src/Server.js:397:9)
Failed to generate resources with Metro, the Metro config may not be using the correct serializer. Ensure the metro.config.js is extending the expo/metro-config and is not overriding the serializer.
(node:47503) ExperimentalWarning: The Fetch API is an experimental feature. This feature could change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
Web .expo/static/render.js ░░░░░░░░░░░░░░░░  0.0% (0/1)
Metro error: Unexpected token < in JSON at position 0

  117 |         const txt = await results.text();
  118 |         try {
> 119 |             return JSON.parse(txt);
      |                         ^
  120 |         } catch (error) {
  121 |             _log.Log.error("Failed to generate resources with Metro, the Metro config may not be using the correct serializer. Ensure the metro.config.js is extending the expo/metro-config and is not overriding the serializer.");
  122 |             debug(txt);

Call Stack
  JSON.parse (<anonymous>)
  MetroBundlerDevServer.getStaticResourcesAsync (node_modules/@expo/cli/build/src/start/server/metro/MetroBundlerDevServer.js:119:25)
  process.processTicksAndRejections (node:internal/process/task_queues)
  async MetroBundlerDevServer.getStaticPageAsync (node_modules/@expo/cli/build/src/start/server/metro/MetroBundlerDevServer.js:149:41)
  async (node_modules/@expo/cli/build/src/start/server/metro/MetroBundlerDevServer.js:261:46)
  ```

**Comment by EloB** [comment]
I also used turborepo but worked for me.

**Comment by dalley8626** [comment]
I've pretty much just ran this command `npx create-expo-app@latest --template tabs@49`

**Comment by dalley8626** [comment]
@EloB are u running on expo-go or expo prebuild?

**Comment by dalley8626** [comment]
i got it working rather than pointing against  `"main": "expo-router/entry",` you should just point to `"main": "index.js",` in package.json. Then in
index.js
```import "expo-router/entry";
```

**Comment by EloB** [comment]
@dalley8626 I already did that so probably why it worked for me :)

**Comment by EloB** [comment]
> try loading mjs with a metro config change
> 
> ```js
> const { getDefaultConfig } = require('expo/metro-config');
> 
> const defaultConfig = getDefaultConfig(__dirname);
> 
> defaultConfig.resolver.sourceExts.push('mjs');
> 
> module.exports = defaultConfig;
> ```

Shouldn't `react-router` have this built in?

**Comment by dalley8626** [comment]
Thanks man, got the same fix as well

**Comment by marklawlor** [comment]
This is an issue with Expo Web, duplicate of https://github.com/expo/expo/issues/23412

**Comment by heroic** [comment]
@marklawlor is this fix planned in a future release?

**Comment by ombogdanJoinToIt** [comment]
I cant use expo prebuild with expo router. why?

**Comment by dalley8626** [comment]
Could u at least show the error log?

**Comment by ombogdanJoinToIt** [comment]
as for me, just my code not reloading after added expo-router

**Comment by Cavdy** [comment]
> try loading mjs with a metro config change
> 
> ```js
> const { getDefaultConfig } = require('expo/metro-config');
> 
> const defaultConfig = getDefaultConfig(__dirname);
> 
> defaultConfig.resolver.sourceExts.push('mjs');
> 
> module.exports = defaultConfig;
> ```

this solved it for me
